rules/dispatch.go

```diff
@@ -387,3 +333,193 @@ func (r *DefaultRuleRenderer) dispatchChains(
 
 	return chains
 }
+
+func (r *DefaultRuleRenderer) dispatchEndPointMarkChains(
+	wlNames []string,
+	hepNames []string,
+	epMarkMapper EndpointMarkMapper,
+	setMarkPfx,
+	wlFromMarkPfx,
+	hepFromMarkPfx,
+	dispatchSetMarkEndpointChainName,
+	dispatchFromMarkEndpointChainName string,
+) []*Chain {
+	names := append(wlNames, hepNames...)
+	// Sort interface names so that rules in the dispatch chain are ordered deterministically.
+	// Otherwise we would reprogram the dispatch chain when there is no real change.
+	sort.Strings(names)
+	log.WithField("ifaceNames", names).Debug("Rendering endpoint mark dispatch chains")
+
+	rootSetMarkRules := make([]Rule, 0)
+	rootFromMarkRules := make([]Rule, 0)
+
+	// start rendering set mark rules.
+	// Since there can be >100 endpoints, putting them in a single list adds some latency to
+	// endpoints that are later in the chain.  To reduce that impact, we build a shallow tree of
+	// chains based on the prefixes of the chains.
+	commonPrefix, prefixes, prefixToNames := r.divideEndpointNamesToPrefixTree(names)
```

Comment:
D'oh, just spotted that this won't work well for a mixed workload and host endpoint chain.  The common prefix of `"caliABC"` and `"eth0"` is `""`, you might need to do host and workload endpoints separately?

Reply:
Yes. Changed to do host and workload endpoints separately. Done.